source/extensions/quic_listeners/quiche/udp_gso_batch_writer.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#pragma GCC diagnostic push
+// QUICHE allows unused parameters.
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+// QUICHE uses offsetof().
+#pragma GCC diagnostic ignored "-Winvalid-offsetof"
+// QUICHE allows ignored qualifiers
+#pragma GCC diagnostic ignored "-Wignored-qualifiers"
+
+// QUICHE doesn't mark override at QuicBatchWriterBase::SupportsReleaseTime()
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winconsistent-missing-override"
+#elif defined(__GNUC__) && __GNUC__ >= 5
+#pragma GCC diagnostic ignored "-Wsuggest-override"
+#endif
+
+#include "quiche/quic/core/batch_writer/quic_gso_batch_writer.h"
+
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
+
+#pragma GCC diagnostic pop
+
+#include "envoy/network/udp_packet_writer_handler.h"
+
+#include "common/protobuf/utility.h"
+#include "common/runtime/runtime_protos.h"
+
+namespace Envoy {
+namespace Quic {
+
+#define UDP_GSO_BATCH_WRITER_STATS(COUNTER, GAUGE, HISTOGRAM)                                      \
+  COUNTER(total_bytes_sent)                                                                        \
+  GAUGE(internal_buffer_size, NeverImport)                                                         \
+  HISTOGRAM(pkts_sent_per_batch, Unspecified)
+
+/**
+ * Wrapper struct for udp gso batch writer stats. @see stats_macros.h
+ */
+struct UdpGsoBatchWriterStats {
+  UDP_GSO_BATCH_WRITER_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT,
+                             GENERATE_HISTOGRAM_STRUCT)
+};
+
+/**
+ * UdpPacketWriter implementation based on quic::QuicGsoBatchWriter to send packets
+ * in batches, using UDP socket's generic segmentation offload(GSO) capability.
+ */
+class UdpGsoBatchWriter : public quic::QuicGsoBatchWriter, public Network::UdpPacketWriter {
```

Comment:
Don't we potentially want to use this outside of QUIC context? Shouldn't this be defined in the core code for use by non-QUIC UDP extensions where the OS supports it? I could imagine this being useful for UDP proxy also?

Reply:
Leaving it here is because it is using QUICHE library. Right now UDP proxy can also use it by linking in the static registered config factory.

Reply:
This means that the user has to compile in QUICHE just to use this functionality. That's probably fine but not clear from the docs. In any case, let's just add a TODO to make this more clear in the docs now or later.

Reply:
Added a TODO to update the Documentation specifying that the user has to compile in QUICHE to use UdpGsoBatchWriter.